### PR TITLE
Graphics Rework

### PR DESCRIPTION
**Pull request by judicaelclair**
**Changes**
- Updated README.md to latest (added Quick Start section with installation instructions)
- Added temporary fix for race conditions between logic and graphics threads (global lock)
- Added `GuiController` and `SimulatorConfig` classes and restructured + ironed out bugs in GUI code.
- `FlatMap` and `SquareMesh` for super fast (both render and load times) and arbitrarily high resolution 2D map visualisation.
- Made OMPL dependency optional
- Added `utility.compatibility` for typing and OMPL detection
- Moved `structures.py` into `structures` folder, which now contains many other structures `Tracked`, `TrackedHeap`, `Heap`, `TrackedList`, `TrackedSet` and factory methods for choosing a tracked or untracked structure based on if graphics with animations is enabled.
- Added `TrackedForest` and `TrackedCyclicGraph` with appropriate factory methods for forests and cyclic graphs.
- Tracked structures are used by the renderer for lazy scene updating (huuuge performance boost).
- Experimental high resolution top-view screenshots.
- Added `Render` with immediate mode style rendering of lines, circles, spheres,...
- Reworked most, if not all, map displays, each having an optimal (or very close to optimal) implementation when tracked data is being used, as well a rudimentary implementation when structure is not tracked.
- Custom heuristics for spheres, circles and line sizes based on map dimensions.
- Essentially reworked entire rendering pipeline.

**Known Issues**
- I advise against using the simulator config window for now, as the hacky fix for race conditions with a lock can lead to a deadlock when starting a new simulation environment.
- Large 2D maps are somewhat sluggish when square colours are being updated. This is because we're editing a really big texture, which Panda3D doesn't like. In the future, partitioning the map will hopefully fix this issue.
- Do not move the agent or goal position while an algorithm is running (wait till animation is completely over), otherwise UB.

GLHF reviewing xD

### CODE REVIEW

**Comment by Nabushika** [comment]
Some stuff about newlines at the end of files, still don't like functions in functions.. 

Loading much improved, still can be slow. Rendering is much faster for 2D maps, 3D maps can still be a little slow.
seems like it works

we should update readme coverage

**Review by Nabushika — APPROVED** [review]

File: src/simulator/views/map/data/flat_map.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        return lbytes, wflbytes, tlbytes
+
+    def render_square(self, p: Point, c: Colour, wfc: Colour) -> None:
+        px, py = int(p[0]), int(p[1])
+        l, wfl, tl = self.__lines[(c, wfc)] if (c, wfc) in self.__lines else self.__create_lines(c, wfc)
+
+        img = self.texture.modify_ram_image()
+        x_offset = px * self.square_size
+        lsize = self.square_size * 4
+        for y in range(py * self.square_size, py * self.square_size + self.wf_line_cnt - 1):
```
Comment (JohnYao93) [inline src/simulator/views/map/data/flat_map.py:152]: don't really know what these are for but you can probably clean these 3 for loops up (or add comments)

File: src/simulator/views/map/display/graph_map_display.py [hunk 2]
```diff
@@ -1,50 +1,218 @@
...
+        if self.__graph.edges_removable:
+            self.__render_lazy_edges_removable(refresh)
+        else:
+            self.__render_lazy_edges_irremovable(refresh)
+
+    def __render_lazy_edges_irremovable(self, refresh: bool) -> None:
+        rv = self.get_renderer_view()
+
+        if refresh:
+            def remove_child(current) -> bool:
```
Comment (JohnYao93) [inline src/simulator/views/map/display/graph_map_display.py:59]: defining functions in other functions like this is probably not the way to go code-style wise. especially as there's another function w the same name that does *almost the same thing in the next function below

File: src/simulator/views/map/display/solid_colour_map_display_pot.py [hunk 3]
```diff
@@ -24,35 +25,36 @@ def __init__(self, services: Services, points, z_index=50, radius=0, custom_map:
...
-                self._root_view.render_pos(Entity(point[0], self.radius), self.color2)
-            elif point[1] < self.small+(self.range*6):
-                self._root_view.render_pos(Entity(point[0], self.radius), self.color1)
-        return True
+        self.pts = set(map(f, points))
+
+        for p in self.pts:
+            self._root_view.cube_requires_update(Entity(p[0], self.radius))
+
+    def update_cube(self, p: Point) -> None:
```
Comment (JohnYao93) [inline src/simulator/views/map/display/solid_colour_map_display_pot.py:47]: is this meant to be a sort of gradient coloring functionality

**Review by JohnYao93 — APPROVED** [review]
looks good overall. Some points of code style can be cleared up

File: src/simulator/views/map/display/solid_colour_map_display_pot.py (hunk 3 above)
Comment (judicaelclair) [reply src/simulator/views/map/display/solid_colour_map_display_pot.py:47]: Don't really know and don't really care xD - never tried to understand the logic

**Pull request merged**